### PR TITLE
Support Consul Connect Envoy Command on Windows

### PR DESCRIPTION
**Pull request by absolutelightning**
### Description

Add support for consul connect envoy command on windows. This PR fixes the comments of PR - https://github.com/hashicorp/consul/pull/15114

### Testing
 *  Built consul.exe from this branch on windows and hosted here - [AWS S3](https://asheshvidyut-bucket.s3.ap-southeast-2.amazonaws.com/consul.zip)
 * Updated the [tutorial](https://developer.hashicorp.com/consul/tutorials/developer-mesh/consul-windows-workloads) and changed the `consul_url.default` value to  [AWS S3](https://asheshvidyut-bucket.s3.ap-southeast-2.amazonaws.com/consul.zip)
 * Followed the steps in the tutorial and verified that everything is working as described.

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -29,6 +29,7 @@ require (
 	github.com/envoyproxy/go-control-plane/xdsmatcher v0.0.0-20230524161521-aaaacbfbe53e
 	github.com/fatih/color v1.14.1
 	github.com/fsnotify/fsnotify v1.6.0
+	github.com/go-mmap/mmap v0.7.0
```
Comment (dhiaayachi) [inline go.mod:32]: If we decide to continue with mmap, I think we should use a better maintained library. https://github.com/edsrzf/mmap-go is an option.

File: .changelog/17694.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+```release-note:feature
+window: support consul connect envoy command on windows
```
Comment (dhiaayachi) [inline .changelog/17694.txt:2]:
```suggestion
Windows: support consul connect envoy command on Windows
```

**Review by dhiaayachi — CHANGES_REQUESTED** [review]
Thank you for working on this @absolutelightning 

I left few comments/suggestions. The only really blocker is the choice of mmap library.

File: go.mod (hunk 1 above)
Comment (absolutelightning) [reply go.mod:32]: This library doesn't have methods like Open, Write, Close and Flush. I am checking how can we use this.

Comment (absolutelightning) [reply go.mod]: @dhiaayachi comments resolved.

File: command/connect/envoy/exec_windows.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	"time"
+
+	"github.com/go-mmap/mmap"
+)
+
+func makeBootstrapTemp(bootstrapJSON []byte) (string, error) {
+	tempFile := filepath.Join(os.TempDir(),
+		fmt.Sprintf("envoy-%x-bootstrap.json", time.Now().UnixNano()+int64(os.Getpid())))
+
+	f, err := mmap.Open(tempFile)
```
[the remark this answers is not in the document]
Comment (absolutelightning) [reply command/connect/envoy/exec_windows.go]: please help me understand we need to use both libraries [npipe](https://github.com/natefinch/npipe) and [mmap](https://github.com/edsrzf/mmap-go) or any one?

[the remark this answers is not in the document]
Comment (dhiaayachi) [reply command/connect/envoy/exec_windows.go]: I think npipe is the best choice as we would like to keep both implementation similar between windows and nix, but if that's not possible we can use mmap.

[the remark this answers is not in the document]
Comment (absolutelightning) [reply command/connect/envoy/exec_windows.go]: This library creates a connection from which we can read lines or can create server. I am not able to understand how we can use it here. Can you please help @dhiaayachi ?

**Comment by absolutelightning** [comment]
@dhiaayachi comments addressed. please review.

File: command/connect/envoy/exec_windows.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	if n < len(bootstrapJSON) {
+		return pipeFile, fmt.Errorf("failed writing boostrap to child STDIN: %s", err)
+	}
+
+	// We can't wait for the process since we need to exec into Envoy before it
+	// will be able to complete so it will be remain as a zombie until Envoy is
+	// killed then will be reaped by the init process (pid 0). This is all a bit
+	// gross but the cleanest workaround I can think of for Envoy 1.10 not
+	// supporting /dev/fd/<fd> config paths any more. So we are done and leaving
+	// the child to run it's course without reaping it.
```
Comment (dhiaayachi) [inline command/connect/envoy/exec_windows.go:59]: I'm not sure I understand the reasoning behind that, why don't we wait for it to be done and is this specific to Windows?

File: command/connect/envoy/exec_windows.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+func execEnvoy(binary string, prefixArgs, suffixArgs []string, bootstrapJSON []byte) error {
+	tempFile, err := makeBootstrapPipe(bootstrapJSON)
+	if err != nil {
+		os.RemoveAll(tempFile)
+		return err
+	}
+	// We don't defer a cleanup since we are about to Exec into Envoy which means
+	// defer will never fire. The child process cleans up for us in the happy
+	// path.
```
Comment (dhiaayachi) [inline command/connect/envoy/exec_windows.go:84]: I'm not sure I understand the reasoning here, why don't we defer the clean up to after the envoy process finish?

**Review by dhiaayachi — COMMENTED** [review]
I left few other questions, I'm not sure I understand some of the comments about not being able to clean up.

File: command/connect/envoy/exec_windows.go (hunk 5 above)
Comment (absolutelightning) [reply command/connect/envoy/exec_windows.go:84]: This log is not specific to windows. Please see - unix version - https://github.com/hashicorp/consul/blob/b0a2e33e0a6c4ec411266dfcef38a3a648fefb50/command/connect/envoy/exec_unix.go#L127C1-L129C10

File: command/connect/envoy/exec_windows.go (hunk 4 above)
Comment (absolutelightning) [reply command/connect/envoy/exec_windows.go:59]: Not specific to Windows. https://github.com/hashicorp/consul/blob/b0a2e33e0a6c4ec411266dfcef38a3a648fefb50/command/connect/envoy/exec_unix.go#L127C1-L129C10

**Review by dhiaayachi — APPROVED** [review]
Thank you for the explanation @absolutelightning 😃 

LGTM!

**Comment by absolutelightning** [comment]
> Thank you for the explanation @absolutelightning 😃
> 
> LGTM!

Thanks.

**Pull request merged**